### PR TITLE
[Feature][API] process definition delete api add check process instance is already running. #3581

### PR DESCRIPTION
**Pull request by zhuangchong**
## What is the purpose of the pull request

The workflow definition deletion function is added to check whether there are running workflow instances according to the workflow definition ID and running status. If there are running workflow instances, they are not allowed to be deleted, and a friendly prompt will be given.    #3581

## Brief change log

New running workflow instance check

ProcessDefinitionServiceImpl.deleteProcessDefinitionById

```
  // check process instances is already running
        List<ProcessInstance> processInstances =  processInstanceMapper.queryByProcessDefineIdAndStatus(processDefinitionId, Constants.NOT_TERMINATED_STATES);
        if(CollectionUtils.isNotEmpty(processInstances)){
            putMsg(result, Status.DELETE_PROCESS_DEFINITION_BY_ID_FAIL,processInstances.size());
            return result;
        }

```

## Verify this pull request

*(Please pick either of the following options)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

ProcessDefinitionControllerTest.testDeleteProcessDefinitionById()

----

##拉取请求的目的是什么

工作流定义删除功能，新增根据工作流定义ID和运行状态，检查是否有正在运行的工作流实例，如果有正在运行的工作流实例，不允许删除，友情提示。

##简要变更日志

新增正在运行工作流实例检查

ProcessDefinitionServiceImpl.deleteProcessDefinitionById

```
  // check process instances is already running
        List<ProcessInstance> processInstances =  processInstanceMapper.queryByProcessDefineIdAndStatus(processDefinitionId, Constants.NOT_TERMINATED_STATES);
        if(CollectionUtils.isNotEmpty(processInstances)){
            putMsg(result, Status.DELETE_PROCESS_DEFINITION_BY_ID_FAIL,processInstances.size());
            return result;
        }

```

##验证此请求

*（请选择以下任一选项）*

此请求已被现有测试覆盖，例如*（请描述测试）*。

ProcessDefinitionControllerTest.testDeleteProcessDefinitionById()

### CODE REVIEW

**Review by qiaozhanwei — APPROVED** [review]
+1

**Comment by zhuangchong** [comment]
> please resolve conflicts
> +1

ok，resolved

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java [hunk 1]
```diff
@@ -488,6 +488,12 @@ private String getResourceIds(ProcessData processData) {
             putMsg(result, Status.PROCESS_DEFINE_STATE_ONLINE, processDefinitionId);
             return result;
         }
+        // check process instances is already running
+        List<ProcessInstance> processInstances =  processInstanceMapper.queryByProcessDefineIdAndStatus(processDefinitionId, Constants.NOT_TERMINATED_STATES);
+        if (CollectionUtils.isNotEmpty(processInstances)) {
+            putMsg(result, Status.DELETE_PROCESS_DEFINITION_BY_ID_FAIL,processInstances.size());
+            return result;
+        }
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java:496]:
```suggestion
        List<ProcessInstance> processInstances =  processInstanceService.queryByProcessDefineIdAndStatus(processDefinitionId, Constants.NOT_TERMINATED_STATES);
        if (CollectionUtils.isNotEmpty(processInstances)) {
            putMsg(result, Status.DELETE_PROCESS_DEFINITION_BY_ID_FAIL,processInstances.size());
            return result;
        }
```

Hi,
In `processDefinitionService` service layer, it is not recommended straightly use `processInstanceMapper` dao layer, this part should be in `processInstanceService`, use `processInstanceService` here.

**Review by yangyichao-mango — CHANGES_REQUESTED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java (hunk 1 above)
Comment (zhuangchong) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ProcessDefinitionServiceImpl.java]: Processinstanceservice has been used. Processinstanceservice should implement the interface. There are contributors who have submitted branches. I have not changed this.

---

已使用processInstanceService ，processInstanceService 应该实现接口，已经有贡献者提交分支，这块我没有改.

**Comment by zhuangchong** [comment]
@dailidong  @qiaozhanwei  done. Please help to review.

**Pull request merged**